frame/balances/src/lib.rs

```diff
@@ -1648,6 +1675,191 @@ impl<T: Config<I>, I: 'static> ReservableCurrency<T::AccountId> for Pallet<T, I>
 	}
 }
 
+impl<T: Config<I>, I: 'static> NamedReservableCurrency<T::AccountId> for Pallet<T, I>  where
+	T::Balance: MaybeSerializeDeserialize + Debug
+{
+	fn reserved_balance_named(id: &ReserveIdentifier, who: &T::AccountId) -> Self::Balance {
+		let reserves = Self::reserves(who);
+		reserves
+			.binary_search_by_key(id, |data| data.id)
+			.map(|index| reserves[index].amount)
+			.unwrap_or_default()
+	}
+
+	/// Move `value` from the free balance from `who` to their reserved balance.
+	///
+	/// Is a no-op if value to be reserved is zero.
+	fn reserve_named(id: &ReserveIdentifier, who: &T::AccountId, value: Self::Balance) -> DispatchResult {
```

Comment:
i'd like another function `ensure_reserved_named`, with the same prototype as this, but which did something roughly equivalent to the following code:

```
fn ensure_reserved_named(id: &ReserveIdentifier, who: &T::AccountId, value: Self::Balance) -> DispatchResult {
  let already = Self:: reserved_balance_named(id, who);
  if already > value { Self::unreserve(already - value); Ok(()) }
  else if value > already { Self::reserve(id, who, value - already) }
  else { Ok(()) }
}
```